### ISSUE TITLE
Very poor image quality results from direct inference

### ISSUE THREAD

**Issue by HumyuShi** [appletea233/EditThinker#4]
Body:

Hi, thank you for the great work and for fully open-sourcing this project. However, when I tried running inference, I found that the output quality was very poor. Also, as the number of edits increases, the images become more severely broken and degraded.

The inference code is shown below. Could you please help me check what might be causing this issue?
`import gc
import json
import os
import re
from typing import Dict

import torch
from transformers import AutoConfig, GenerationConfig
from PIL import Image
from diffusers import DiffusionPipeline, QwenImageEditPipeline
from transformers import AutoModelForVision2Seq, AutoProcessor

def main():
    os.makedirs(OUTPUT_DIR, exist_ok=True)

    prompt_template = load_prompt_template(PROMPT_TEMPLATE_PATH)

    torch.manual_seed(SEED)
    _patch_generation_config()

    # ---- Load T2I pipeline ----
    t2i_pipe = DiffusionPipeline.from_pretrained(
        QWEN_T2I_MODEL,
        torch_dtype=DTYPE,
        trust_remote_code=True,
    )
    t2i_pipe = t2i_pipe.to(DEVICE)

    # ---- Generate initial image ----
    generator = torch.Generator(device=DEVICE).manual_seed(SEED)
    with torch.inference_mode():
        t2i_out = t2i_pipe(
            prompt=T2I_PROMPT,
            true_cfg_scale=GUIDANCE_SCALE,
            num_inference_steps=STEPS,
            width=WIDTH,
            height=HEIGHT,
            generator=generator,
        ).images[0]

    original_image = t2i_out
    if DEVICE == "cuda":
        t2i_pipe.to("cpu")
    del t2i_pipe
    gc.collect()
    if DEVICE == "cuda":
        torch.cuda.empty_cache()
        torch.cuda.ipc_collect()
    original_path = os.path.join(OUTPUT_DIR, "t2i_0.png")
    save_image(original_image, original_path)
    print(f"Saved initial T2I image: {original_path}")

    # del t2i_pipe
    # torch.cuda.empty_cache()
    
    rewritten_prompt = EDIT_INSTRUCTION
    current_image = original_image

    for turn in range(MAX_TURNS):
        # ---- Load Edit pipeline (per turn) ----
        edit_pipe = QwenImageEditPipeline.from_pretrained(
            QWEN_EDIT_MODEL,
            torch_dtype=DTYPE,
        )
        if DEVICE == "cuda" and EDIT_USE_CPU_OFFLOAD:
            try:
                edit_pipe.enable_model_cpu_offload()
            except Exception:
                edit_pipe = edit_pipe.to(DEVICE)
        else:
            edit_pipe = edit_pipe.to(DEVICE)

        # ---- Edit ----
        generator = torch.Generator(device=DEVICE).manual_seed(SEED)
        # with torch.inference_mode():
        edit_out = edit_pipe(
            image=current_image,
            prompt=rewritten_prompt,
            true_cfg_scale=GUIDANCE_SCALE,
            num_inference_steps=STEPS,
            generator=generator,
        ).images[0]

        if DEVICE == "cuda":
            edit_pipe.to("cpu")
        del edit_pipe
        gc.collect()
        if DEVICE == "cuda":
            torch.cuda.empty_cache()
            torch.cuda.ipc_collect()

        edited_image = edit_out
        edited_path = os.path.join(OUTPUT_DIR, f"edit_{turn+1}.png")
        save_image(edited_image, edited_path)
        print(f"Turn {turn+1}: saved edited image: {edited_path}")

        
        # ---- Load EditThinker (reasoning) ----
        processor = AutoProcessor.from_pretrained(EDITTHINKER_MODEL, trust_remote_code=True)
        reason_model = AutoModelForVision2Seq.from_pretrained(
            EDITTHINKER_MODEL,
            torch_dtype=DTYPE,
            device_map="auto",
            trust_remote_code=True,
            config=AutoConfig.from_pretrained(EDITTHINKER_MODEL, trust_remote_code=True),
        )
        reason_model.eval()

        # ---- Reasoning ----
        prompt_text = prompt_template.replace("{original_instruction}", EDIT_INSTRUCTION).replace(
            "{rewritten_prompt}", rewritten_prompt
        )

        messages = [
            {
                "role": "user",
                "content": [
                    {"type": "image", "image": original_image},
                    {"type": "image", "image": edited_image},
                    {"type": "text", "text": prompt_text},
                ],
            }
        ]

        text = processor.apply_chat_template(messages, tokenize=False, add_generation_prompt=True)
        inputs = processor(
            text=[text],
            images=[original_image, edited_image],
            padding=True,
            return_tensors="pt",
        ).to(reason_model.device)

        with torch.inference_mode():
            output_ids = reason_model.generate(**inputs, max_new_tokens=512)
        gen_text = processor.batch_decode(
            output_ids[:, inputs.input_ids.shape[1]:], skip_special_tokens=True
        )[0]

        del reason_model
        del processor
        gc.collect()
        if DEVICE == "cuda":
            torch.cuda.empty_cache()
            torch.cuda.ipc_collect()

        eval_result = parse_editthinker_output(gen_text)
        print(
            f"Turn {turn+1} eval score: {eval_result.get('score', {})}, satisfied={eval_result['is_satisfied']}"
        )

        if eval_result["is_satisfied"]:
            break

        if eval_result["new_rewritten_prompt"]:
            rewritten_prompt = eval_result["new_rewritten_prompt"]
        current_image = edited_image
        with open('instruction.txt', 'a') as tgt:
            tgt.write(f'turn {turn} instruction : {rewritten_prompt}\n')
            tgt.flush()

    meta_path = os.path.join(OUTPUT_DIR, "run_meta.json")
    with open(meta_path, "w", encoding="utf-8") as f:
        json.dump(
            {
                "t2i_prompt": T2I_PROMPT,
                "edit_instruction": EDIT_INSTRUCTION,
                "final_prompt": rewritten_prompt,
                "seed": SEED,
                "turns": MAX_TURNS,
            },
            f,
            ensure_ascii=False,
            indent=2,
        )
    print(f"Saved metadata: {meta_path}")

if __name__ == "__main__":
    main()
`
T2I model is QwenImage and edit with QwenImageEdit

**Comment by appletea233** [comment]
Thanks a lot for reporting this, and also for sharing the code in detail.

We checked the cause, and the main issue is very likely **CFG was not actually enabled** in your current inference setup.

In the `diffusers` pipeline version you are using, the edit pipeline does **not automatically provide a default `negative_prompt`**. As a result, even though you pass `true_cfg_scale=GUIDANCE_SCALE`, the model may still effectively run **without classifier-free guidance**, which means the behavior is close to **`cfg = 1`**. This usually leads to noticeably worse editing quality, and in multi-turn editing it can make the image drift and degrade much more severely after each step.

So in short:

* your current pipeline is likely running **without real CFG**
* previous broken / blurry / drifting results are therefore consistent with **effective `cfg=1`**
* once an explicit `negative_prompt` is provided, CFG should be properly activated

A likely fix is to pass an explicit negative prompt when calling `QwenImageEditPipeline`, for example:

```python
edit_out = edit_pipe(
    image=current_image,
    prompt=rewritten_prompt,
    negative_prompt="",
    true_cfg_scale=GUIDANCE_SCALE,
    num_inference_steps=STEPS,
    generator=generator,
).images[0]
```

**Comment by HumyuShi** [comment]
> Thanks a lot for reporting this, and also for sharing the code in detail.
> 
> We checked the cause, and the main issue is very likely **CFG was not actually enabled** in your current inference setup.
> 
> In the `diffusers` pipeline version you are using, the edit pipeline does **not automatically provide a default `negative_prompt`**. As a result, even though you pass `true_cfg_scale=GUIDANCE_SCALE`, the model may still effectively run **without classifier-free guidance**, which means the behavior is close to **`cfg = 1`**. This usually leads to noticeably worse editing quality, and in multi-turn editing it can make the image drift and degrade much more severely after each step.
> 
> So in short:
> 
> * your current pipeline is likely running **without real CFG**
> * previous broken / blurry / drifting results are therefore consistent with **effective `cfg=1`**
> * once an explicit `negative_prompt` is provided, CFG should be properly activated
> 
> A likely fix is to pass an explicit negative prompt when calling `QwenImageEditPipeline`, for example:
> 
> edit_out = edit_pipe(
>     image=current_image,
>     prompt=rewritten_prompt,
>     negative_prompt="",
>     true_cfg_scale=GUIDANCE_SCALE,
>     num_inference_steps=STEPS,
>     generator=generator,
> ).images[0]

<img width="1306" height="456" alt="Image" src="https://github.com/user-attachments/assets/6529c7af-e925-4bec-b91f-61182cc45642" />

Thanks for your response. I added a negative prompt, and that fixed the issue.
